agent/wsclient/client.go

```diff
@@ -285,3 +280,76 @@ func (cs *ClientServerImpl) handleMessage(data []byte) {
 		log.Info("No handler for message type", "type", typeStr)
 	}
 }
+
+// websocketConn establishes a connection to the given URL, respecting any proxy configuration in the environment.
+// A standard proxying setup involves setting the following environment variables
+// (may be listed in /etc/ecs/ecs.config if using ecs-init):
+// HTTP_PROXY=http://<your-proxy>/ # HTTPS_PROXY may be set instead or additionally
+// NO_PROXY=169.254.169.254,/var/run/docker.sock # Directly connect to metadata service and docker socket
+func (cs *ClientServerImpl) websocketConn(parsedURL *url.URL, request *http.Request) (net.Conn, error) {
+	proxyURL, err := http.ProxyFromEnvironment(request)
+	if err != nil {
+		return nil, err
+	}
+
+	// url.Host might not have the port, but tls.Dial needs it
+	targetHost := parsedURL.Host
+	if !strings.Contains(targetHost, ":") {
+		targetHost += ":443"
+	}
+	targetHostname, _ , err := net.SplitHostPort(targetHost)
+	if err != nil {
+		return nil, err
+	}
+
+	tlsConfig := tls.Config{ServerName: targetHostname, InsecureSkipVerify: cs.AcceptInvalidCert}
+	timeoutDialer := &net.Dialer{Timeout: wsConnectTimeout}
+
+	if proxyURL == nil {
+		// directly connect
+		log.Info("Creating poll dialer", "host", parsedURL.Host)
+		return tls.DialWithDialer(timeoutDialer, "tcp", targetHost, &tlsConfig)
+	}
+
+	// connect via proxy
+	log.Info("Creating poll dialer", "proxy", proxyURL.Host, "host", parsedURL.Host)
+	plainConn, err := timeoutDialer.Dial("tcp", proxyURL.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	// TLS over an HTTP proxy via CONNECT taken from: https://golang.org/src/net/http/transport.go
+	connectReq := &http.Request{
+		Method: "CONNECT",
+		URL: &url.URL{Opaque: targetHost},
+		Host: targetHost,
+		Header: make(http.Header),
+	}
+
+	if proxyUser := proxyURL.User; proxyUser != nil {
+		username := proxyUser.Username()
+		password, _ := proxyUser.Password()
+		auth := base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
+		connectReq.Header.Set("Proxy-Authorization", "Basic " + auth)
+	}
+
+	connectReq.Write(plainConn)
+
+	// Read response.
+	// Okay to use and discard buffered reader here, because
+	// TLS server will not speak until spoken to.
+	br := bufio.NewReader(plainConn)
+	resp, err := http.ReadResponse(br, connectReq)
+	if err != nil {
+		plainConn.Close()
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		plainConn.Close()
+		return nil, errors.New(strconv.Itoa(resp.StatusCode))
```

Comment:
Switched from Split to Itoa.  Just a note: this is now creating an error with the status code, not the reason (you get "401" instead of "Unauthorized").